posts/2022-01-03-kryo-pitfalls.adoc

```diff
@@ -0,0 +1,289 @@
+This benchmark uses a similar class model from our IoT devices, defining a `Message` containing a list of `Signal`.
+
+Finally, https://github.com/twitter/chill[Twitter Chill] provides this fair warning:
+
+____
+Serialization compatibility is NOT guaranteed between releases, and for this reason, we don't recommend
+using it for long-term storage. Serialization is highly dependent on scala version compatibility and on
+the underlying Kryo serializers, which take different approaches to compatibility.
+____
+
+**Conclusion:** Registering classes makes Kryo more rigid, but it's also faster, which, if we combine with
+including the class in the payload, we avoid headaches at read time. If our desire to keep this reading safety
+but be more flexible, we end up in the worst speed/space quadrant. This is not to say that it's useless, just
+that you have to consider your use case carefully, as sometimes some speed/space penalty can be paid.
+
+Kryo also offers, like other serialization libraries, Backwards and Forwards compatibility, just not by
+default, and we have to put extra thought into another tradeoff. _There is no free lunch_.
+
+
+== Pitfall #2: Composition
+
+While serializing, both writer and reader must agree on the schema before they can start interacting.
+Different libraries handle this differently, you can read about this on
+https://martin.kleppmann.com/2012/12/05/schema-evolution-in-avro-protocol-buffers-thrift.html[Martin Kleppmann's blogpost].
+On Kryo, the way this agreement is done is that both writer and schema must agree on the identifiers (e.g.
+the `0c` ID from the previous section) they use for each of the classes involved (unless registration is
+turned off and the class name is serialized inside the payload), which could also lead to problems.
+
+Imagine we have tree components: a Device, a Processor, and a Consumer. Each send information to the next
+using Kryo (let's ignore the medium for now).
+
+----
++----------------+      +----------------+      +----------------+
+|     Device     |----->|    Processor   |----->|    Consumer    |
++----------------+      +----------------+      +----------------+
+----
+
+And that the _Device_ declares the classes from the previous section: `Message`, `Signal`, etc. These are then
+aggregated by the _Processor_ which defines its own classes: `Event` and `Notification`. Which are then consumed
+by the _Consumer_ (this latter component doesn't care about the first set of types).
+
+When registering classes, we cannot reuse IDs, otherwise we'd run into conflict or disagreement. For example,
+imagine that both the _Device_ and the _Processor_ uses ID 10 for `Message` and that the _Processor_ use ID 11 for `Event`.
+But after a few months of running in production, the _Device_ decides to use ID 11, which breaks the _Processor_'s
+operation.
+
+This, together with the performance requirement of using https://github.com/EsotericSoftware/kryo#pooling[Pooling],
```

Comment:
Just wondering; You mention "performance requirement to use `Pooling`", that's probably because for performance, multi-threading was required which led to making things thread-safe, which required either using `ThreadLocal` or `Pooling`. Would it be a good idea to mention this, as it draws attention on the fact that Kryo isn't thread-safe?

Reply:
Yes, I modified the sentence, it wasn't entirely clear why I was bringing up Pooling